### PR TITLE
Optimize cuvs-lucene CAGRA_HNSW index build; add example

### PR DESCRIPTION
**Pull request by jamxia155**
This PR replaces [cuVS-Lucene #195](https://github.com/NVIDIA/cuvs-lucene/pull/195) as cuVS-Lucene has been merged into cuVS.

This PR builds on the out-of-core host-streaming refactor from #2476 by @nvzm123. This PR subsumes #2476 to avoid stacking the PRs.

On top of #2476, this PR adds the following.

**Improvements**

- Native flat buffering: stream vectors directly into a native host matrix during indexing instead of buffering them as a heap List<float[]>, cutting peak host memory from ~2x to ~1x and eliminating the per-vector matrix-assembly copy. Opt-in; requires all input vectors to be indexed in the original order, i.e. not supported for sorted, merged, or filtered index segments; not currently supported for segments with quantized fields.
- Parallelized CAGRA-to-HNSW graph conversion: materializing the CAGRA adjacency into on-heap NeighborArrays was a serial per-node loop; parallelized under the existing `writerThreads` knob.
- Parallelized level-0 HNSW graph serialization: level-0 (all N nodes) is now delta/VInt-encoded in parallel across memory-bounded waves of threads, then concatenated to the `IndexOutput` in node order according to on-disk format in the serial path.

**New example: OptimizedCagraHnswBuildExample**

A reference pattern for building a large accelerated HNSW index with all ingest- and build-side optimizations, showcasing:

- Streaming, prefetched, bounded-memory ingestion: open the source file once, read it front-to-back in large sequential chunks; hold at most two chunks in memory; fill the next chunk on a background thread while the ingest thread drains the current one, hiding disk read behind indexing; unpack into a caller-reused float[] allocation (no per-vector allocation, safe because Lucene copies the value eagerly inside `addDocument`).
- Native flat buffering sized exactly to each segment's vector count via `withNumInputVectors`, avoiding the heap-buffered assembly copy.
- Overlapping multi-segment build:
    - Default: K single-segment passes appended to one directory; peak host memory is one slice (N/K).
    - Overlapped: a bounded pool (`PIPELINE_DEPTH`) builds segments concurrently into their own directories with the GPU commit serialized on a semaphore (ingest overlaps a prior segment's GPU commit), then combines the finished per-segment indexes by hardlinking their files into the final directory via `HardlinkCopyDirectoryWrapper` + `addIndexes` (no bulk copy of vector data). Peak host memory is up to `PIPELINE_DEPTH * N / K`.
- An `enableRMMAsyncMemory()` call (with a note that it must not be used with CPU-only codecs) to opt-in to RMM-managed memory resources, plus exposing the primary tuning knobs (withMaxConn, withBeamWidth, withCuvsDistanceType, withWriterThreads).

### CODE REVIEW

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java [hunk 1]
```diff
@@ -235,56 +243,149 @@ private static CuVSMatrix buildCagraGraphForSubset(
...
+                    return null;
+                  }));
         }
-        // Write the size after duplicates are removed
-        vectorIndex.writeVInt(actualSize);
-        // Write de-duplicated neighbors
-        for (int i = 0; i < actualSize; i++) {
-          vectorIndex.writeVInt(scratch[i]);
+        for (Future<?> f : futures) {
+          f.get();
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java]: If one of these tasks throws  InterruptedException or an ExecutionException  the remaining tasks in that wave keep running to completion even after this method exists. I think it might be prudent to call shutdownNow() in case of an exception so we interrupt all other threads and call shutdown() only in the case of a success.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/NativeFlatVectorsWriter.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+ *       https://github.com/apache/lucene/blob/releases/lucene/10.2.0/lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99FlatVectorsWriter.java
+ * </ul>
+ *
+ * <p><b>On a Lucene upgrade:</b> re-verify this class against the new version's two files above. The
+ * constants are package-private in Lucene (hence re-declared here), and the file format is validated
+ * on read via {@code CodecUtil.checkIndexHeader} (codec name + version range) plus the fixed meta
+ * layout — so if the new version bumps {@code VERSION_CURRENT}, renames a codec, or changes the meta
+ * field order, files written here will be silently incompatible and the stock
+ * {@code Lucene99FlatVectorsReader} will reject or misread them. Concretely:
+ *
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/NativeFlatVectorsWriter.java:52]: This doesn't sound quite right to me. I think the guidance should be the opposite: once this is written, we should *not* touch these constants.

Whatever we write with the current `Lucene99FlatVectorsFormat` v0 layout stays readable through the last Lucene 11.x release, since Lucene guarantees back-compat for one major version and `backward_codecs` keeps the v0 read path alive. So we can keep upgrading `lucene-core` while leaving this serialization exactly as it is.

The constants aren't really ours to sync anyway. The codec names and extensions are a handshake with the reader. `checkIndexHeader` string-matches the name and the extensions pick the files, so Lucene can't change them without breaking every existing index. And `VERSION_CURRENT` is the version *we emit*, not Lucene's. Bumping it while leaving `writeField`/`writeMeta` untouched is the single change most likely to actually break things: we'd stamp a new version number onto old-layout bytes, and the reader would misparse them silently rather than reject them.

When we do decide to move to a newer format, it won't be a constants update - we'll have to re-examine the whole binary serialization. So it's safer to leave the constants pinned until we're ready to do that.

What this paragraph should say instead, I think: on upgrade the only thing to verify is that the stock reader still accepts what we write - which `TestNativeFlatVectorsWriterRoundTrip` already does, against whatever reader is on the classpath.

File: java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterFormatConstants.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * {@code lucene-core} jar actually on the classpath (its manifest {@code Specification-Version},
+ * not a parse of {@code pom.xml}'s text), so a Lucene upgrade fails this test immediately rather
+ * than silently writing {@code .vec}/{@code .vemf} files the stock
+ * {@code Lucene99FlatVectorsReader} may no longer read correctly.
+ *
+ * <p>This intentionally fires on <em>any</em> version change, not just ones that actually alter
+ * the format — per the class javadoc, every upgrade needs a manual re-verification pass against
+ * the new version's {@code Lucene99FlatVectorsFormat}/{@code Lucene99FlatVectorsWriter} sources.
+ */
+public class TestNativeFlatVectorsWriterFormatConstants {
```
Comment (imotov) [inline java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterFormatConstants.java]: I don't think this test based on my previous comment on NativeFlatVectorsWriter.

File: java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterRoundTrip.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                + " Lucene99FlatVectorsReader",
+            dataset[Integer.parseInt(id)],
+            roundTripped,
+            0f);
+        seen++;
+      }
+      assertEquals("did not visit every vector", numDocs, seen);
+    }
+  }
+}
```
Comment (imotov) [inline java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterRoundTrip.java:279]: I think we could use a couple more tests here:
- The same round-trip test with 10000 docs, but with only a random 8000 of them carrying the vector field. Cardinality above `IndexedDISI.MAX_ARRAY_LENGTH` (4095) selects the DENSE encoding path.
- The same round-trip test with 1000 docs and a random 800 carrying the field, which stays under that threshold and takes the SPARSE path instead.
- A test with more than one vector field, or a tweak to one of the existing tests, so we cover the boundary between consecutive per-field records in `.vemf`.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java [hunk 5]
```diff
@@ -235,56 +243,149 @@ private static CuVSMatrix buildCagraGraphForSubset(
...
+          final int slot = t;
+          if (subStart >= subEnd) {
             continue;
           }
-          scratch[actualSize++] = nnodes[i] - nnodes[i - 1];
+          futures.add(
+              pool.submit(
+                  () -> {
+                    ByteBuffersDataOutput buffer = new ByteBuffersDataOutput();
+                    int[] scratch = new int[graph.maxConn() * 2];
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java]: `graph.maxConn()` is not trivial. Might makes sense to extract it outside of the loop.

File: java/cuvs-lucene/examples/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+    // disable RAM-based flushing so nothing flushes before commit, and forbid merges. This is what
+    // makes native flat buffering valid. Order matters: enable the doc-count trigger BEFORE
+    // disabling
+    // the RAM trigger, since Lucene rejects a config where both are disabled at once.
+    IndexWriterConfig config =
+        new IndexWriterConfig()
+            .setCodec(codec)
+            .setUseCompoundFile(false)
+            .setMaxBufferedDocs(Math.max(2, size + 1))
+            .setRAMBufferSizeMB(IndexWriterConfig.DISABLE_AUTO_FLUSH)
```
Comment (imotov) [inline java/cuvs-lucene/examples/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java]: This is a pretty strong requirement that might not be acceptable by Solr and Elasticsearch.

**Review by imotov — COMMENTED** [review]
Sorry for the delay. I wanted to better understand the logic and explore a couple of alternatives.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java (hunk 1 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java]: Updated as suggested. Thanks.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/NativeFlatVectorsWriter.java (hunk 2 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/NativeFlatVectorsWriter.java:52]: You're right that `Lucene99FlatVectorsFormat` should retain compatibility for the major version. Revised the tests/documentation and removed unnecessary tripwire tests.

File: java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterFormatConstants.java (hunk 3 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterFormatConstants.java]: Removed.

File: java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterRoundTrip.java (hunk 4 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/test/java/com/nvidia/cuvs/lucene/TestNativeFlatVectorsWriterRoundTrip.java:279]: Added tests as suggested. Thanks!

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java (hunk 5 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/AcceleratedHNSWUtils.java]: Great spot! Extracted outside loop here and in one more instance.

File: java/cuvs-lucene/examples/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java (hunk 6 above)
Comment (jamxia155) [reply java/cuvs-lucene/examples/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java]: Updated documentation to highlight the prerequisites for using the native flat buffering optimization. Now framed as an opportunistic optimization that is constrained by multiple pre-conditions.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/CagraIndexParamsFactory.java [hunk 7]
```diff
@@ -18,6 +25,121 @@ public class CagraIndexParamsFactory {
 
   private CagraIndexParamsFactory() {}
 
+  /**
+   * Translation of the internal logic found in the constructor of {@code struct ivf_pq_params} in
+   * cuVS's {@code cpp/include/cuvs/neighbors/ivf_pq.hpp}.
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/CagraIndexParamsFactory.java]: Ok, I am really confused now. We just removed this in https://github.com/NVIDIA/cuvs-lucene/pull/177/changes#diff-b2d360e425d107d219d1a101fb52f6fdcb1643229dfda96c40dae7e01904c899L31 in order to delegate this logic to cuVS, we do we resurrect it again here?

File: examples/java/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java [hunk 8]
```diff
@@ -0,0 +1,669 @@
+/*
```
Comment (cjnolet) [inline examples/java/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java:1]: Let's please move this `examples/java`. cuvs lucene is a library.

File: java/cuvs-lucene/examples/pom.xml [hunk 9]
```diff
@@ -92,6 +92,11 @@
         <artifactId>lucene-backward-codecs</artifactId>
         <version>10.2.0</version>
     </dependency>
+    <dependency>
```
Comment (cjnolet) [inline java/cuvs-lucene/examples/pom.xml]: `examples/java` is going to need to build its own jar (just how the examples for the other languages build their own artifacts)

File: examples/java/README.md [hunk 10]
```diff
@@ -32,3 +32,17 @@ To run the Index and Search on GPU example do:
 ```sh
 mvn clean install && java -Djava.util.logging.config.file=src/main/resources/logging.properties -cp target/examples-26.10.0-jar-with-merged-services.jar com.nvidia.cuvs.lucene.examples.IndexAndSearchonGPUExample
 ```
+
+To run the optimized CAGRA-HNSW build example (reference pattern for efficiently building an
```
Comment (cjnolet) [inline examples/java/README.md:36]: Please move to examples/java

File: examples/java/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java (hunk 8 above)
Comment (jamxia155) [reply examples/java/src/main/java/com/nvidia/cuvs/lucene/examples/OptimizedCagraHnswBuildExample.java:1]: Moved.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java [hunk 11]
```diff
@@ -93,6 +109,13 @@ public Lucene99AcceleratedHNSWVectorsWriter(
     this.flatVectorsWriter = flatVectorsWriter;
     this.infoStream = state.infoStream;
     this.acceleratedHNSWParams = acceleratedHNSWParams;
+    this.numInputVectors = acceleratedHNSWParams.getNumInputVectors();
+    this.nativeMode = numInputVectors > 0;
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java]: When `nativeMode` here is `true` this writer doesn't actually follow the contract of KnnVectorsWriter faithfully, which means it shouldn't be used by anything but CagraHnswBulkIndexWriter. So, I think instead of branching inside more universal Lucene99AcceleratedHNSWVectorWriterit should be a separate class implementing KnnVectorsWriter, preferably package private is it can only be instantiated within the limits of CagraHnswBulkIndexWriter. I would also remove getNumInputVectors() from acceleratedHNSWParams while at it. I don't think we need to do it as part of this PR, so if you don't have time to get to it, please open an issue for it and assign it to me, so we don't lose it.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/CagraHnswBulkIndexWriter.java [hunk 12]
```diff
@@ -0,0 +1,660 @@
...
+      buildOverlapped(fbinPath, config, callback, chunkSizeMB, slices, dim);
+    } else {
+      try (FbinVectorSource source = new FbinVectorSource(fbinPath, chunkSizeMB)) {
+        buildSequential(source, config, callback, slices);
+      }
+    }
+  }
+
+  private static void ensureRMMEnabled() {
+    if (RMM_ENABLED.compareAndSet(false, true)) {
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/CagraHnswBulkIndexWriter.java]: That looks like a race condition. The second and the following threads that will reach this point are not going to wait for the first thread to actually set RMM. Instead of setting it here, I would rather check that it is already set and bail out with an error if it is not.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FieldWriter.java [hunk 13]
```diff
@@ -23,18 +25,64 @@ public class FieldWriter extends KnnFieldVectorsWriter<Object> {
...
+  private final int numInputVectors;
+
+  private final boolean nativeBuffering;
+  private final CuVSMatrix.Builder<CuVSHostMatrix> hostMatrixBuilder;
+  private final DocsWithFieldSet nativeDocsWithField;
+  private int nativeCount;
+  private CuVSHostMatrix builtMatrix;
+
   @SuppressWarnings("unchecked")
   public FieldWriter(
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FieldWriter.java:32]: I think this should be also a separate NativeFieldWriter it is just too different from the "normal" one.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java [hunk 14]
```diff
@@ -213,6 +280,37 @@ public void flush(int maxDoc, DocMap sortMap) throws IOException {
...
+              + " numInputVectors vectors were added; see AcceleratedHNSWParams.Builder"
+              + "#withNumInputVectors.");
+    }
+    FieldInfo fieldInfo = fieldData.fieldInfo();
+    try {
+      CuVSHostMatrix dataset = fieldData.getHostMatrix();
+      nativeFlat.writeField(fieldInfo, dataset, maxDoc, fieldData.getDocsWithFieldSet());
+      writeFieldInternal(fieldInfo, dataset);
+    } finally {
+      fieldData.releaseNativeBuffer();
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java]: This might not be the most reliable place to release the native buffer. It looks like if we have wrong numInputVectors or sort key or any other error situation, we never reach this point.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FbinVectorSource.java [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    current = next;
+  }
+
+  @Override
+  public void get(int index, float[] dst) throws IOException {
+    if (index < 0 || index >= windowSize) {
+      throw new IndexOutOfBoundsException(
+          "Index " + index + " out of bounds [0, " + windowSize + ")");
+    }
+    if (index < nextExpectedRelative) {
```
Comment (imotov) [inline java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FbinVectorSource.java:172]: VectorSource docs say "index must be greater than **or equal** to the index passed to the previous call, if any". That seems a bit inconsistent.

**Review by imotov — COMMENTED** [review]
Looks much better! Thanks! Left a few comments.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FbinVectorSource.java (hunk 15 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FbinVectorSource.java:172]: Fixed docs (code was correct).

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java (hunk 14 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java]: Add a `releaseAllNativeBuffers()` backstop to `close()`.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FieldWriter.java (hunk 13 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/FieldWriter.java:32]: Split as suggested.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java (hunk 11 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/Lucene99AcceleratedHNSWVectorsWriter.java]: Split as suggested.

File: java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/CagraHnswBulkIndexWriter.java (hunk 12 above)
Comment (jamxia155) [reply java/cuvs-lucene/src/main/java/com/nvidia/cuvs/lucene/CagraHnswBulkIndexWriter.java]: This is indeed a race condition and even worse, it goes against our existing convention of leaving RMM enablement as an opt-in in application code. Sticking to the convention naturally resolves this as the typical pattern is to enable RMM before any GPU resources are created.

Also updated the documentation for `enableRMM*` and `resetRMMPooledMemory` methods to clearly advise against self-enable/reset within classes.

File: examples/java/README.md (hunk 10 above)
Comment (imotov) [reply examples/java/README.md:36]: @cjnolet there is no examples/java at the moment - https://github.com/NVIDIA/cuvs/tree/main/examples. Java examples currently live in two places: [`java/examples`](https://github.com/NVIDIA/cuvs/tree/main/java/examples) for generic cuVS examples and [`java/cuvs-lucene/examples`](https://github.com/NVIDIA/cuvs/tree/main/java/cuvs-lucene/examples) for Lucene-specific examples. I have it on my [to-do list](https://github.com/NVIDIA/cuvs/issues/2288) to move them into /examples/java in the near future, but I don't think we should do it as part of this commit since it would (at least temporarily) introduce a third place where Java examples can be found.

**Review by jameslamb — APPROVED** [review]
Approving for CI/packaging codeowners groups (CI/packaging changes look minimal and non-controversial).

**Review by imotov — APPROVED** [review]
LGTM, just waiting for #2543 to be merge so we don't spread examples all over the place.
